exp/controllers/gcpmanagedcontrolplane_controller.go

```diff
@@ -29,33 +49,163 @@ import (
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultedLoopTimeout(r.ReconcileTimeout))
+	defer cancel()
+
+	log := ctrl.LoggerFrom(ctx)
+
+	// Get the control plane instance
+	gcpManagedControlPlane := &infrav1exp.GCPManagedControlPlane{}
+	if err := r.Client.Get(ctx, req.NamespacedName, gcpManagedControlPlane); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	// Get the cluster
+	cluster, err := util.GetOwnerCluster(ctx, r.Client, gcpManagedControlPlane.ObjectMeta)
+	if err != nil {
+		log.Error(err, "Failed to retrieve owner Cluster from the API Server")
+		return ctrl.Result{}, err
+	}
+	if cluster == nil {
+		log.Info("Cluster Controller has not yet set OwnerRef")
+		return ctrl.Result{}, nil
+	}
+
+	if annotations.IsPaused(cluster, gcpManagedControlPlane) {
+		log.Info("Reconciliation is paused for this object")
+		return ctrl.Result{}, nil
+	}
+
+	// Get the managed cluster
+	managedCluster := &infrav1exp.GCPManagedCluster{}
+	key := client.ObjectKey{
+		Namespace: gcpManagedControlPlane.Namespace,
+		Name:      cluster.Spec.InfrastructureRef.Name,
+	}
+	if err := r.Client.Get(ctx, key, managedCluster); err != nil || managedCluster == nil {
+		log.Error(err, "Failed to retrieve GCPManagedCluster from the API Server")
+		return ctrl.Result{}, err
+	}
+
+	managedControlPlaneScope, err := scope.NewManagedControlPlaneScope(ctx, scope.ManagedControlPlaneScopeParams{
+		Client:                 r.Client,
+		Cluster:                cluster,
+		GCPManagedCluster:      managedCluster,
+		GCPManagedControlPlane: gcpManagedControlPlane,
+	})
+	if err != nil {
+		return ctrl.Result{}, errors.Errorf("failed to create scope: %+v", err)
+	}
 
-	// TODO(user): your logic here
+	// Always close the scope when exiting this function so we can persist any GCPMachine changes.
+	defer func() {
+		if err := managedControlPlaneScope.Close(); err != nil && reterr == nil {
+			reterr = err
+		}
+	}()
+
+	// Handle deleted clusters
+	if !gcpManagedControlPlane.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, managedControlPlaneScope)
+	}
+
+	// Handle non-deleted clusters
+	return r.reconcile(ctx, managedControlPlaneScope)
+}
+
+func (r *GCPManagedControlPlaneReconciler) reconcile(ctx context.Context, managedControlPlaneScope *scope.ManagedControlPlaneScope) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+	log.Info("Reconciling GCPManagedControlPlane")
+
+	controllerutil.AddFinalizer(managedControlPlaneScope.GCPManagedControlPlane, infrav1exp.ManagedControlPlaneFinalizer)
+	if err := managedControlPlaneScope.PatchObject(); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if !managedControlPlaneScope.GCPManagedCluster.Status.Ready {
+		log.Info("GCPManagedCluster not ready yet, retry later")
+		return ctrl.Result{RequeueAfter: 2 * time.Minute}, nil
+	}
+
+	reconcilers := []cloud.ReconcilerWithResult{
```

Comment:
We could reuse this across reconcile and reconcileDelete

Reply:
could you explain a bit what do you mean by this? `reconcile` and `reconcileDelete` both use `ReconcilerWithResult` so i don't quite understand it. Thanks!

Reply:
We can chat about this at a later date. For the time being lets resolve this